tests/test_lzip_decoder.c

```diff
@@ -0,0 +1,454 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+/// \file       test_lzip_decoder.c
+/// \brief      Tests decoding lzip data
+//
+//  Author:     Jia Tan
+//
+//  This file has been put into the public domain.
+//  You can do whatever you want with this file.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "tests.h"
+
+#ifdef HAVE_LZIP_DECODER
+
+// Memlimit large enough to pass all of the test files
+#define MEMLIMIT (1U << 20)
+#define DECODE_CHUNK_SIZE 1024
+
+
+// Avoiding using the ASCII string because of potential EBCDIC system
+// compatibility. The string is: "Hello\nWorld!\n"
+static const char hello_world[] = {'H', 'e', 'l', 'l', 'o', '\n',
+			'W', 'o', 'r', 'l', 'd', '!', '\n', 0};
+// The string is "Trailing garbage\n"
+static const char trailing_garbage[] = {'T', 'r', 'a', 'i', 'l', 'i',
+		'n', 'g', ' ', 'g', 'a', 'r', 'b', 'a', 'g', 'e', '\n', 0};
+
+
+// Helper function to decode a good file with no flags and plenty
+// high memlimit
+static void
+basic_lzip_decode(const char *src, const uint8_t *expected_result,
+		size_t expected_result_size) {
+	size_t file_size;
+	uint8_t *data = tuktest_file_from_srcdir(src, &file_size);
+
+	lzma_stream strm = LZMA_STREAM_INIT;
+	assert_lzma_ret(lzma_lzip_decoder(&strm, MEMLIMIT, 0), LZMA_OK);
+
+	uint8_t *output_buffer = tuktest_malloc(expected_result_size);
+
+	strm.next_in = data;
+	strm.next_out = output_buffer;
+	strm.avail_out = expected_result_size;
+
+	// Feed 1 byte at a time to the decoder to look for any bugs
+	// when switching between decoding sequences
+	lzma_ret ret = LZMA_OK;
+	while (ret == LZMA_OK) {
+		strm.avail_in = 1;
+		ret = lzma_code(&strm, LZMA_RUN);
+	}
+
+	assert_lzma_ret(ret, LZMA_STREAM_END);
+
+	assert_uint_eq(strm.total_in, file_size);
+	assert_uint_eq(strm.total_out, expected_result_size);
+	assert_array_eq(output_buffer, expected_result,
+			expected_result_size);
+	lzma_end(&strm);
+}
+
+
+static void
+test_options(void)
+{
+	// Test NULL stream
+	assert_lzma_ret(lzma_lzip_decoder(NULL, MEMLIMIT, 0),
+			LZMA_PROG_ERROR);
+
+	// Test invalid flags
+	lzma_stream strm = LZMA_STREAM_INIT;
+	assert_lzma_ret(lzma_lzip_decoder(&strm, MEMLIMIT, UINT32_MAX),
+			LZMA_OPTIONS_ERROR);
+	// Memlimit tests are done elsewhere
+}
+
+
+static void
+test_v0_decode(void) {
+	// This tests if liblzma can decode lzip version 0 files.
+	// lzip 1.17 and older can decompress this, but lzip 1.18
```

Comment:
I wonder if it's not useful to repeat the version numbers of the other tool here as things might change and they aren't relevant for liblzma testing. They are described in the tests/files/README in detail already.